// harvest/dist/api/estimates.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
class EstimatesAPI {
    constructor(harvest) {
        this.baseUrl = '/v2/estimates';
        this.harvest = harvest;
    }
    get(id) {
        return this.harvest.request('GET', `${this.baseUrl}/${id}`);
    }
    list(query = {}) {
        return this.harvest.request('GET', this.baseUrl, query);
    }
    create(data) {
        return this.harvest.request('POST', this.baseUrl, data);
    }
    update(id, data) {
        return this.harvest.request('PATCH', `${this.baseUrl}/${id}`, data);
    }
    delete(id) {
        return this.harvest.request('DELETE', `${this.baseUrl}/${id}`);
    }
    send(id) {
        return this.harvest.request('POST', `${this.baseUrl}/${id}/messages`, {
            event_type: 'send'
        });
    }
    accept(id) {
        return this.harvest.request('POST', `${this.baseUrl}/${id}/messages`, {
            event_type: 'accept'
        });
    }
    decline(id) {
        return this.harvest.request('POST', `${this.baseUrl}/${id}/messages`, {
            event_type: 'decline'
        });
    }
    reopen(id) {
        return this.harvest.request('POST', `${this.baseUrl}/${id}/messages`, {
            event_type: 're-open'
        });
    }
}
exports.EstimatesAPI = EstimatesAPI;
