// harvest/dist/api/invoices.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
class InvoicesAPI {
    constructor(harvest) {
        this.baseUrl = '/v2/invoices';
        this.harvest = harvest;
    }
    get(id) {
        return this.harvest.request('GET', `${this.baseUrl}/${id}`);
    }
    list(query = {}) {
        return this.harvest.request('GET', this.baseUrl, query);
    }
    create(data) {
        return this.harvest.request('POST', this.baseUrl, data);
    }
    update(id, data) {
        return this.harvest.request('PATCH', `${this.baseUrl}/${id}`, data);
    }
    delete(id) {
        return this.harvest.request('DELETE', `${this.baseUrl}/${id}`);
    }
    draft(id) {
        return this.harvest.request('POST', `${this.baseUrl}/${id}/messages`, {
            event_type: 'draft'
        });
    }
    send(id) {
        return this.harvest.request('POST', `${this.baseUrl}/${id}/messages`, {
            event_type: 'send'
        });
    }
    close(id) {
        return this.harvest.request('POST', `${this.baseUrl}/${id}/messages`, {
            event_type: 'close'
        });
    }
    reopen(id) {
        return this.harvest.request('POST', `${this.baseUrl}/${id}/messages`, {
            event_type: 're-open'
        });
    }
}
exports.InvoicesAPI = InvoicesAPI;
